### PR TITLE
fix(polys): fix DMP.cancel for GF(p) with Flint

### LINKED ISSUE #26804: Poly.cancel fails over GF with ground types flint
This is after gh-25940
```python
In [1]: Poly(x, modulus=11).cancel(Poly(x, modulus=11))
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[1], line 1
----> 1 Poly(x, modulus=11).cancel(Poly(x, modulus=11))

File ~/current/active/sympy/sympy/polys/polytools.py:3890, in Poly.cancel(f, g, include)
   3887 dom, per, F, G = f._unify(g)
   3889 if hasattr(F, 'cancel'):
-> 3890     result = F.cancel(G, include=include)
   3891 else:  # pragma: no cover
   3892     raise OperationNotSupported(f, 'cancel')

File ~/current/active/sympy/sympy/polys/polyclasses.py:849, in DMP.cancel(f, g, include)
    847     return F._cancel_include(G)
    848 else:
--> 849     return F._cancel(G)

File ~/current/active/sympy/sympy/polys/polyclasses.py:2143, in DUP_Flint._cancel(f, g)
   2140 """Cancel common factors in a rational function ``f/g``. """
   2141 # Think carefully about how to handle denominators and coefficient
   2142 # canonicalisation if more domains are permitted...
-> 2143 assert f.dom == g.dom in (ZZ, QQ)
   2145 if f.dom.is_QQ:
   2146     cG, F = f.clear_denoms()

AssertionError: 
```

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Fixes gh-26804

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/polys/polyclasses.py [hunk 1]
```diff
@@ -2138,19 +2138,28 @@ def _lcm(f, g):
...
             cF, G = g.clear_denoms()
         else:
-            cG, F = f.dom.one, f
-            cF, G = g.dom.one, g
+            cG, F = R.one, f
+            cF, G = R.one, g
 
-        cH = cF.gcd(cG)
-        cF, cG = cF // cH, cG // cH
+        if R.is_ZZ or R.is_QQ:
```
Comment (sylee957) [inline sympy/polys/polyclasses.py:2160]: Is it necessary to check `is_ZZ or is_QQ` here? `is_FiniteField` is narrowed down from above

Comment (oscarbenjamin) [reply sympy/polys/polyclasses.py:2160]: Oh, good point

**Comment by sylee957** [comment]
It may be useful to check if field can follow same procedure as `QQ` and can use `clear_denoms` 
But I found some other error in the implementation

```python3
from sympy import *

x = Symbol('x')
Poly(x**2 - 1, modulus=2).clear_denoms()
```
`AttributeError: 'flint.types.nmod_poly.nmod_poly' object has no attribute 'denom'`

**Comment by oscarbenjamin** [comment]
> It may be useful to check if field can follow same procedure as `QQ` and can use `clear_denoms`

The part that is different for `GF(p)` is that we don't normalise the sign of the leading coefficients.

> But I found some other error in the implementation

Thanks. I've fixed that and been through all of the implementation. I found some places where other improvements could be made e.g. to avoid converting to `DMP_Python`.

I found and fixed one other bug:
```python
In [1]: Poly(x, x, modulus=10**100)
---------------------------------------------------------------------------
TypeError
```

**Comment by sylee957** [comment]
I notice another error with factoring over non-prime modulus.
```
(sympy-dev-py3.11) sylee957@DESKTOP-2LJUD4V:~/documents/sympy-dev$ python
Python 3.11.2 (main, Mar 21 2023, 16:28:24) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> x = Symbol('x')
>>> Poly(x**2 - 1, modulus=4).factor_list()
Flint exception (Impossible inverse):
    Cannot invert modulo 2*2
Aborted
```

Although this should be an error, 
however, I think that the situation is different that this leads to crash of python
that is not possible to recover with try-catch statement.

**Comment by oscarbenjamin** [comment]
Actually the simplest solution for now would be to not use flint for non-prime modulus.

**Comment by oscarbenjamin** [comment]
I believe I have fixed the issues mentioned above. This is ready for review.

**Comment by sylee957** [comment]
Thanks

**Pull request merged**